lib/spack/spack/solver/asp.py

```diff
@@ -918,15 +967,18 @@ def key_fn(version):
 
         for weight, declared_version in enumerate(most_to_least_preferred):
             self.gen.fact(
-                fn.version_declared(
-                    pkg.name, declared_version.version, weight, str(declared_version.origin)
+                fn.facts(
```

Comment:
Done in https://github.com/spack/spack/pull/38447/commits/c721ebffb8f7be241e1898e6046b5f33618c34e7

> I know this is somewhat hypocritical of me because [ .. ]

I don't think it is. Code benefits from improved naming, so having pushed a name that is too generic in the past isn't a good reason to let another one get in right now We need to change `attr` in a following PR :slightly_smiling_face: